### ISSUE TITLE
Testing pre-trained models on a custom video

### ISSUE THREAD

**Issue by JVD9kh96** [ubicomplab/rPPG-Toolbox#191]
Hello, thanks for your awesom repository. I noticed that it is possible to train and test various number of models on the mentioned dataset, but I don't know how can I run the pre-trained models on a custom video for obtaining rPPG signals. Any tips?
Thanks in advance

**Comment by yahskapar** [comment]
We currently don't support inference on custom videos, but this should be quite straightforward to do on your own fork of this repo if you take a look at some of the config examples in [`configs/infer_configs`](https://github.com/ubicomplab/rPPG-Toolbox/tree/main/configs/infer_configs) that leverage a pre-trained model to test on another dataset. In your case, you'd have to make a custom dataloader for your custom dataset based on an existing dataloader, such as [`UBFCrPPGLoader.py`](https://github.com/ubicomplab/rPPG-Toolbox/blob/main/dataset/data_loader/UBFCrPPGLoader.py), update `main.py` under [here](https://github.com/ubicomplab/rPPG-Toolbox/blob/main/main.py#L211) to utilize that dataloader for your custom dataset, and then create your own inference config file for your custom dataset.

Give that a shot and feel free to post back here if you run into any trouble.

**Comment by JVD9kh96** [comment]
> We currently don't support inference on custom videos, but this should be quite straightforward to do on your own fork of this repo if you take a look at some of the config examples in [`configs/infer_configs`](https://github.com/ubicomplab/rPPG-Toolbox/tree/main/configs/infer_configs) that leverage a pre-trained model to test on another dataset. In your case, you'd have to make a custom dataloader for your custom dataset based on an existing dataloader, such as [`UBFCrPPGLoader.py`](https://github.com/ubicomplab/rPPG-Toolbox/blob/main/dataset/data_loader/UBFCrPPGLoader.py), update `main.py` under [here](https://github.com/ubicomplab/rPPG-Toolbox/blob/main/main.py#L211) to utilize that dataloader for your custom dataset, and then create your own inference config file for your custom dataset.
> 
> Give that a shot and feel free to post back here if you run into any trouble.


Thanks for your kind reply. I implemented a custom data generator based on your explanations and it was really helpful. My implementation could be found [here ](https://github.com/JVD9kh96/rPPG-Toolbox/tree/dev) and updated README file based on it. I tried to stick as much as possible to your implementation, so that if you found it useful maybe it could help others who want a solution for custom videos. 

I had two more questions. The first one is that I need the networks to estimate the rPPG signals, because I need it for a project. If I take the raw output of the models in [here ](https://github.com/JVD9kh96/rPPG-Toolbox/blob/6542b8591b045f4aeb1de2cc2ca785e1fe20b39d/evaluation/metrics.py#L63), this is the rPPG signal, am I right? 

The second one is that, since there are no labels for a custom video, I made the data loader [generate pseudo-labels](https://github.com/JVD9kh96/rPPG-Toolbox/blob/07425d68ba0ca689ccd334d0a0dd5dbc5fcc22ad/dataset/data_loader/CustomTestLoader.py#L57). After the [evaluation ](https://github.com/JVD9kh96/rPPG-Toolbox/blob/07425d68ba0ca689ccd334d0a0dd5dbc5fcc22ad/inference.py#L153)with metrics calculation, I get the following results: 

FFT MAE (FFT Label): 8.203125 +/- 5.800485314420897
FFT RMSE (FFT Label): 11.600970628841795 +/- 95.16421218971784
FFT MAPE (FFT Label): 10.0 +/- 7.0710678118654755
/content/evaluation/metrics.py:152: RuntimeWarning: invalid value encountered in double_scalars
  standard_error = np.sqrt((1 - correlation_coefficient**2) / (num_test_samples - 2))
FFT Pearson (FFT Label): -1.0 +/- nan
FFT SNR (FFT Label): -7.7797272173904455 +/- 0.4013820937968198 (dB)


Since they are pseudo-labels, I don't know getting `nan` pearson metric is good or not! Are these metrics trustworthy for pseudo-labels? 

Thanks for your time

**Comment by yahskapar** [comment]
Nice work with your custom implementation!

> I had two more questions. The first one is that I need the networks to estimate the rPPG signals, because I need it for a project. If I take the raw output of the models in [here ](https://github.com/JVD9kh96/rPPG-Toolbox/blob/6542b8591b045f4aeb1de2cc2ca785e1fe20b39d/evaluation/metrics.py#L63), this is the rPPG signal, am I right?

Yes, but I'd also take a look at how that signal is post-processed in [ `metrics.py`](https://github.com/JVD9kh96/rPPG-Toolbox/blob/6542b8591b045f4aeb1de2cc2ca785e1fe20b39d/evaluation/metrics.py), and eventually [`post_process.py`](https://github.com/JVD9kh96/rPPG-Toolbox/blob/6542b8591b045f4aeb1de2cc2ca785e1fe20b39d/evaluation/post_process.py). You may need to do a similar kind of post-processing in your own project.

> Since they are pseudo-labels, I don't know getting nan pearson metric is good or not! Are these metrics trustworthy for pseudo-labels?

Those metrics should still be good to use with pseudo-labels, and the reason your are getting a `nan` is most likely because you have only one or two custom videos that makes `num_test_samples` lead to an invalid value in this context (from either dividing by 0 or eventually taking the sqrt of a negative quantity that leads to an imaginary number) as the runtime warning indicates. Try this out with three custom videos - you should notice the `nan` disappear at that point.

Note that it's very possible the generated pseudo-label is unhelpful to begin with, which I personally would expect to be the case with certain kinds of videos, for example ones that have heavy motion and that a unsupervised PPG estimation method such as POS would most likely struggle with. There are probably quite a few other factors to consider there, maybe @girishvn can comment a bit based on his own experience with using pseudo-labels and given he contributed that functionality to this repo.

**Comment by JVD9kh96** [comment]
> Nice work with your custom implementation!
> 
> > I had two more questions. The first one is that I need the networks to estimate the rPPG signals, because I need it for a project. If I take the raw output of the models in [here ](https://github.com/JVD9kh96/rPPG-Toolbox/blob/6542b8591b045f4aeb1de2cc2ca785e1fe20b39d/evaluation/metrics.py#L63), this is the rPPG signal, am I right?
> 
> Yes, but I'd also take a look at how that signal is post-processed in [ `metrics.py`](https://github.com/JVD9kh96/rPPG-Toolbox/blob/6542b8591b045f4aeb1de2cc2ca785e1fe20b39d/evaluation/metrics.py), and eventually [`post_process.py`](https://github.com/JVD9kh96/rPPG-Toolbox/blob/6542b8591b045f4aeb1de2cc2ca785e1fe20b39d/evaluation/post_process.py). You may need to do a similar kind of post-processing in your own project.
> 
> > Since they are pseudo-labels, I don't know getting nan pearson metric is good or not! Are these metrics trustworthy for pseudo-labels?
> 
> Those metrics should still be good to use with pseudo-labels, and the reason your are getting a `nan` is most likely because you have only one or two custom videos that makes `num_test_samples` lead to an invalid value in this context (from either dividing by 0 or eventually taking the sqrt of a negative quantity that leads to an imaginary number) as the runtime warning indicates. Try this out with three or more custom videos - you should notice the `nan` disappear at that point.
> 
> Note that it's very possible the generated pseudo-label is unhelpful to begin with, which I personally would expect to be the case with certain kinds of videos, for example ones that have heavy motion and that an unsupervised PPG estimation method such as POS would most likely struggle with. There are probably quite a few other factors to consider there, maybe @girishvn can comment a bit based on his own experience with using pseudo-labels and given he contributed that functionality to this repo.

Thanks for you kind help, I appreciate that. 

I think you are right, I double-ckecked the code and turned out I might need to apply `_detrend` and `band_pass` filter to the prediction, based on the frequency of sampling `Fs` and `diff_flag` based on the information in the corresponding config file. So, I made [process_raw_prediction ](https://github.com/JVD9kh96/rPPG-Toolbox/blob/e54e3f85712af3dd4c76ed7d9b63f9fb4ebc0ec9/evaluation/post_process.py#L99) function in this regard, and apply it to the network predictions where I want to [save the predictions](https://github.com/JVD9kh96/rPPG-Toolbox/blob/4d0cd27d4e7989685e66045ea3b1b8689183ccd2/evaluation/metrics.py#L122).  


And about pseudo-labels, the nan problem was solved when I run the inference on multiple videos. Thank you very much!

**Comment by yahskapar** [comment]
No problem - happy to hear things are working now. I'll go ahead and close this thread now that it seems the issue has been resolved, but please feel free to re-open if you feel otherwise, or make a new issue if you run into any bugs or anything that you need clarification on.
